packages/grafana-data/src/types/variables.ts

```diff
@@ -0,0 +1,93 @@
+import { ComponentType } from 'react';
+import { Observable } from 'rxjs';
+
+import {
+  DataQuery,
+  DataQueryRequest,
+  DataQueryResponse,
+  DataSourceApi,
+  DataSourceJsonData,
+  DataSourceOptionsType,
+  DataSourceQueryType,
+  QueryEditorProps,
+} from './datasource';
+
+/**
+ * Defines new variable support
+ *
+ * Use one of the properties standard | custom | datasource to define the kind of query editor and query support you want to use in the Variable system
+ *
+ * @alpha -- experimental
+ */
+export interface VariableSupport<
+  DSType extends DataSourceApi<TQuery, TOptions>,
+  TQuery extends DataQuery = DataSourceQueryType<DSType>,
+  TOptions extends DataSourceJsonData = DataSourceOptionsType<DSType>,
+  VariableQuery extends DataQuery = any
+> {
+  /**
+   * Use this property if you want to use the standard query editor supplied in the Variable system
+   */
+  standard?: StandardVariableSupport<DSType, TQuery, TOptions>;
+
+  /**
+   * Use this property if you want to use a custom query and query editor
+   */
+  custom?: CustomVariableSupport<DSType, VariableQuery, TQuery, TOptions>;
+
+  /**
+   * Use this property if you want to use the same query editor as the data source
+   */
+  datasource?: DataSourceVariableSupport<DSType, TQuery, TOptions>;
```

Comment:
Do we need this? You can just supply the same editor as a custom editor right? That should do the same thing doesn't it?

Reply:
I see what you mean but not really, because for the data source variable support we'll use the `query` function from the data source so you only need to declare the editor. An improvement I can see is that we could probably resolve the editor from the data source plugin instead.